README.md
# KIND
Kubernetes in Docker helper.
`kind_helper` command creates kubernetes cluster with local registry and NGINX ingress controller.

## Preparation
### macOS
Install Docker Desktop for Mac
```
brew install homebrew/cask/docker
```

Install `kubectl` command line and overwrite the symlinks created by Docker Desktop for Mac.
```
brew install kubernetes-cli
brew link --overwrite kubernetes-cli
```

Install `kind` command line.
```
brew install kind
```
### Linux
Install `docker` by following [Install Docker Engine on Debian](https://docs.docker.com/engine/install/debian) documentation.

Install `kubectl` by following [Install and Set Up kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl) documentation.

Install `kind` by following [Quick Start](https://kind.sigs.k8s.io/docs/user/quick-start/#installation) documentation.

### Windows
TODO

## Installation
### macOS/Linux
```
curl -fsSL https://raw.githubusercontent.com/humansriot/kind/master/install.sh | bash
```
### Windows
TODO

Validate the installation finished successfully.
```
kind_helper doctor
```

## Update
In order to update `kind_helper` command.
```
kind_helper update
```

## Usage
Create cluster with default configuration.
```
kind_helper create
```

Create cluster with given configuration.
```
kind_helper create config.yaml
```

Create cluster with given configuration and change name.
```
kind_helper create config.yaml name
```

Test cluster
```
kind_helper test
```


CHANGELOG.md
# Changelog

## [Unreleased]
### Fixed
- Replace kind-registry by localhost in ConfigMap to work with Tilt
 
## [0.2.1] - 2021-01-06
### Fixed
- Don't display entering/leaving directory (from make)

## [0.2.0] - 2021-01-06
### Added
- `doctor` command
### Fixed
- Fallback with Git clone via HTTP during installation

## [0.1.0] - 2021-01-06
### Added
- `test` command

## [0.0.1] - 2021-01-06
- Initial release

[Unreleased]: https://github.com/humansriot/kind/compare/0.2.1...HEAD
[0.2.1]: https://github.com/humansriot/kind/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/humansriot/kind/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/humansriot/kind/compare/0.0.1...0.1.0
[0.0.1]: https://github.com/humansriot/kind/releases/tag/0.0.1


test_ingress.sh
#!/usr/bin/env bash
set -euo pipefail

wait_until_curl() {
  request=$1

  printf "%s: " "$request"
  until curl --output /dev/null --silent --head --fail "$request"; do
    printf '.'
    sleep 1
  done
  echo "OK"
}

# Apply example-ingress
echo
echo '$ kubectl apply -f https://kind.sigs.k8s.io/examples/ingress/usage.yaml'
kubectl apply -f https://kind.sigs.k8s.io/examples/ingress/usage.yaml

# Wait until pods are ready to process requests
echo
echo "@ Waiting for pods are ready to process requests"
kubectl wait \
  --for=condition=ready pod \
  --selector=app=foo \
  --timeout=90s
kubectl wait \
  --for=condition=ready pod \
  --selector=app=bar \
  --timeout=90s

# Test ingress
echo
echo '$ kubectl describe ingress example-ingress'
kubectl describe ingress example-ingress
echo
wait_until_curl http://localhost/foo
wait_until_curl http://localhost/bar

# Delete example-ingress
echo
echo '$ kubectl delete -f https://kind.sigs.k8s.io/examples/ingress/usage.yaml'
kubectl delete -f https://kind.sigs.k8s.io/examples/ingress/usage.yaml


test_registry.sh
#!/usr/bin/env bash
set -euo pipefail

reg_host=localhost
reg_port='5000'
reg_url=$reg_host:$reg_port

# Push http-echo to local registry
echo
echo "@ Pull, tag and push http-echo image"
HTTP_ECHO_VERSION=0.2.3
docker pull hashicorp/http-echo:$HTTP_ECHO_VERSION
docker tag hashicorp/http-echo:$HTTP_ECHO_VERSION $reg_url/http-echo:$HTTP_ECHO_VERSION
docker push $reg_url/http-echo:$HTTP_ECHO_VERSION

# Deploy http-echo from local registry
echo
echo "$ kubectl apply -f -"
APP=echo-app
cat <<EOF | kubectl apply -f -
kind: Pod
apiVersion: v1
metadata:
  name: $APP
  labels:
    app: foo
spec:
  containers:
  - name: $APP
    image: $reg_url/http-echo:$HTTP_ECHO_VERSION
    args:
    - "-text=echo"
EOF

# Wait until pod is ready
echo
echo "@ Waiting for pods are ready to process requests"
kubectl wait \
  --for=condition=ready \
  pod $APP \
  --timeout=90s

# Delete POD
echo
echo "$ kubectl delete pod $APP"
kubectl delete pod $APP


kind_create.sh
#!/usr/bin/env bash
set -euo pipefail

KIND_SCRIPT_PATH=$(dirname "$0")
KIND_SCRIPT_PATH=$(cd "$KIND_SCRIPT_PATH" && pwd)
KIND_CLUSTER_CONFIG=${1:-$KIND_SCRIPT_PATH/cluster.yaml}
KIND_CLUSTER_NAME="${2:-kind}"

# Create registry container unless it already exists
reg_name='kind-registry'
reg_host=${reg_name}
reg_port='5000'
running="$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)"
if [ "${running}" != 'true' ]; then
  docker run \
    -d --restart=always -p "${reg_port}:5000" --name "${reg_name}" \
    registry:2
fi

# Create a cluster with the local registry enabled in containerd
TMP=$(mktemp)
cat "$KIND_CLUSTER_CONFIG" > "$TMP"
cat >> "$TMP" <<EOF
containerdConfigPatches:
- |-
  [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:${reg_port}"]
    endpoint = ["http://${reg_host}:${reg_port}"]
EOF
cat "$TMP"
kind create cluster --name="$KIND_CLUSTER_NAME" --config="$TMP"
rm "$TMP"

# Connect the registry to the cluster network
# (the network may already be connected)
docker network connect "kind" "${reg_name}" || true

# Document the local registry
# https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: local-registry-hosting
  namespace: kube-public
data:
  localRegistryHosting.v1: |
    host: "${reg_host}:${reg_port}"
    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
EOF

# Deploy Ingress NGINX controller
NGINX_INGRESS_CONTROLLER_VERSION=0.43.0
kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v${NGINX_INGRESS_CONTROLLER_VERSION}/deploy/static/provider/kind/deploy.yaml
kubectl wait --namespace ingress-nginx \
  --for condition=available \
  deployment ingress-nginx-controller \
  --timeout=90s
kubectl wait --namespace ingress-nginx \
  --for=condition=ready \
  pod --selector=app.kubernetes.io/component=controller \
  --timeout=90s


kind_helper.sh
#!/usr/bin/env bash
set -euo pipefail

function follow_links() {
  file="$1"
  while [ -h "$file" ]; do
    # On Mac OS, readlink -f doesn't work.
    file="$(readlink "$file")"
  done
  echo "$file"
}

HELPER_NAME=$(basename "$0")
HELPER_PATH=$(dirname "$(follow_links "$0")")

function check_command() {
  local name
  local command
  name=$1
  command=$2

  if $command &> /dev/null; then
    echo
    echo "$ $command"
    $command
  else
    echo "$name is missing"
  fi
}

function doctor() {
  check_command docker "docker --version"
  check_command kubectl "kubectl version"
  check_command kind "kind version"
  check_command kind_helper "kind_helper version"
}

function create_cluster() {
  local cluster_config
  local cluster_name
  cluster_config=$1
  cluster_name=$2

  # Create registry container unless it already exists
  reg_name='kind-registry'
  reg_host=${reg_name}
  reg_port='5000'
  running="$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)"
  if [ "${running}" != 'true' ]; then
    docker run \
      -d --restart=always -p "${reg_port}:5000" --name "${reg_name}" \
      registry:2
  fi

  # Create a cluster with the local registry enabled in containerd
  TMP=$(mktemp)
  cat "$cluster_config" > "$TMP"
  cat >> "$TMP" <<EOF
containerdConfigPatches:
- |-
  [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:${reg_port}"]
    endpoint = ["http://${reg_host}:${reg_port}"]
EOF
  cat "$TMP"
  kind create cluster --name="$cluster_name" --config="$TMP"
  rm "$TMP"

  # Connect the registry to the cluster network
  # (the network may already be connected)
  docker network connect "kind" "${reg_name}" || true

  # Document the local registry
  # https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry
  cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: local-registry-hosting
  namespace: kube-public
data:
  localRegistryHosting.v1: |
    host: "localhost:${reg_port}"
    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
EOF

  # Deploy Ingress NGINX controller
  NGINX_INGRESS_CONTROLLER_VERSION=0.43.0
  kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v${NGINX_INGRESS_CONTROLLER_VERSION}/deploy/static/provider/kind/deploy.yaml
  kubectl wait --namespace ingress-nginx \
    --for condition=available \
    deployment ingress-nginx-controller \
    --timeout=90s
  kubectl wait --namespace ingress-nginx \
    --for=condition=ready \
    pod --selector=app.kubernetes.io/component=controller \
    --timeout=90s
}

function help() {
  echo "${HELPER_NAME}: ${VERSION} on ${UNAME}"
  echo ""
  echo "Usage: ${HELPER_NAME} <command>"
  echo ""
  echo "Available commands:"
  echo "  create [<config>] [<name>]    Create cluster (optional: <config> file, <name> cluster name)"
  echo "  test                          Test cluster"
  echo ""
  echo "  changelog     Show changelog"
  echo "  version       Update ${HELPER_NAME} command"
  echo "  update        Update ${HELPER_NAME} command"
}

UNAME=$(uname -s)
CMD=${1:-""}
MAKE="make -C $HELPER_PATH --no-print-directory"
VERSION=$($MAKE version)

case ${CMD} in
  "changelog")
    $MAKE changelog
    ;;
  "version")
    $MAKE version
    ;;
  "update")
    $MAKE update
    ;;
  "doctor")
    doctor
    ;;
  "create")
    shift
    create_cluster "${1:-$HELPER_PATH/cluster.yaml}" "${2:-kind}"
    ;;
  "test")
    "$HELPER_PATH"/test_ingress.sh
    "$HELPER_PATH"/test_registry.sh
    ;;
  *)
    help
    ;;
esac
